org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/generators/xml2dtd/XML2DTDGeneratorTest.java

```diff
@@ -0,0 +1,208 @@
+		String xml = "<note>\r\n" + //
+				"	<to>Tove</to>\r\n" + //
+				"	<from>Jani</from>\r\n" + //
+				"	<heading>Reminder</heading>\r\n" + //
+				"	<body>Don't forget me this weekend!</body>\r\n" + //
+				"</note>";
+		String dtd = "<!ELEMENT note (to, from, heading, body)>" + lineSeparator() + //
+				"<!ELEMENT to (#PCDATA)>" + lineSeparator() + //
+				"<!ELEMENT from (#PCDATA)>" + lineSeparator() + //
+				"<!ELEMENT heading (#PCDATA)>" + lineSeparator() + //
+				"<!ELEMENT body (#PCDATA)>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void dtdWithAttr() throws IOException {
+		String xml = "<note version=\"1.2\" />";
+		String dtd = "<!ELEMENT note EMPTY>" + lineSeparator() + //
+				"<!ATTLIST note version CDATA #IMPLIED>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void empty() throws IOException {
+		String xml = "<note />";
+		String dtd = "<!ELEMENT note EMPTY>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void onlyCharacterContent() throws IOException {
+		String xml = "<note>ABCD</note>";
+		String dtd = "<!ELEMENT note (#PCDATA)>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void onlyElementChildren() throws IOException {
+		String xml = "<note><from>ABCD</from></note>";
+		String dtd = "<!ELEMENT note (from)>" + lineSeparator() + //
+				"<!ELEMENT from (#PCDATA)>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void characterContentAndElementChildren() throws IOException {
+		String xml = "<note>ABCD<from>ABCD</from></note>";
+		String dtd = "<!ELEMENT note (#PCDATA|from)*>" + lineSeparator() + //
+				"<!ELEMENT from (#PCDATA)>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void characterContentAndSeveralElementChildren() throws IOException {
+		String xml = "<note><from>ABCD</from><to></to></note>";
+		String dtd = "<!ELEMENT note (from, to)>" + lineSeparator() + //
+				"<!ELEMENT from (#PCDATA)>" + lineSeparator() + //
+				"<!ELEMENT to EMPTY>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void threeLevel() {
+		String xml = "<a>\r\n" + //
+				"	<b>\r\n" + //
+				"		<c/>\r\n" + //
+				"	</b>\r\n" + //
+				"</a>";
+		String dtd = "<!ELEMENT a (b)>" + lineSeparator() + //
+				"<!ELEMENT b (c)>" + lineSeparator() + //
+				"<!ELEMENT c EMPTY>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void severalSameElement() {
+		String xml = "<a>\r\n" + //
+				"	<b>\r\n" + //
+				"		<c />\r\n" + //
+				"		<c />\r\n" + //
+				"	</b>\r\n" + //
+				"	<b />\r\n" + //
+				"	<b>\r\n" + //
+				"		<c />\r\n" + //
+				"	</b>\r\n" + //
+				"</a>";
+		String dtd = "<!ELEMENT a (b+)>" + lineSeparator() + //
+				"<!ELEMENT b (c+)>" + lineSeparator() + //

```

Comment:
I think this should be `<!ELEMENT b (c*)>`, since there is a `<b />`

Reply:
Indeed there were a bug and it should be fixed.